collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java

```diff
@@ -77,205 +51,84 @@ private JSONObject getModifiedPrefs(Collection<String> keys) throws JSONExceptio
 
         JSONObject generalPrefsJson = settingsJson.getJSONObject("general");
         JSONObject adminPrefsJson = settingsJson.getJSONObject("admin");
 
-        Collection<String> allKeys = getAllKeys();
-        for (String key : allKeys) {
-
-            if (key.equals(KEY_ADMIN_PW) && adminPrefsJson.has(key)) {
-                adminEditor.putString(key, adminPrefsJson.getString(key));
-                adminEditor.apply();
-
-                // skip further checking
-                continue;
-            }
+        for (String key : getAllGeneralKeys()) {
 
             if (generalPrefsJson.has(key)) {
-                try {
-                    boolean bool = generalPrefsJson.getBoolean(key);
-                    editor.putBoolean(key, bool);
-                } catch (Exception e) {
-                    try {
-                        String string = generalPrefsJson.getString(key);
-                        editor.putString(key, string);
-                    } catch (Exception e1) {
-                        editor.putLong(key, generalPrefsJson.getLong(key));
-                    }
-                }
+                Object value = generalPrefsJson.get(key);
+                GeneralSharedPreferences.getInstance().save(key, value);
             } else {
-                try {
-                    String stringValue = getStringValue(key);
-                    editor.putString(key, getDefaultStringValue(key));
-                } catch (ClassCastException e) {
-                    try {
-                        boolean booleanValue = getBooleanValue(key);
-                        editor.putBoolean(key, getDefaultBooleanValue(key));
-                    } catch (ClassCastException e1) {
-                        long longValue = getLongValue(key);
-                        editor.putLong(key, getDefaultLongValue(key));
-                    }
-                }
+                GeneralSharedPreferences.getInstance().reset(key);
             }
         }
-        editor.apply();
 
-        for (String key : ALL_KEYS) {
+        for (String key : getAllAdminKeys()) {
+
             if (adminPrefsJson.has(key)) {
-                adminEditor.putBoolean(key, false);
+                Object value = adminPrefsJson.get(key);
+                AdminSharedPreferences.getInstance().save(key, value);
             } else {
-                adminEditor.putBoolean(key, true);
+                AdminSharedPreferences.getInstance().reset(key);
             }
-            adminEditor.apply();
         }
 
         //settings import confirmation toast
         ToastUtils.showLongToast(context.getString(R.string.successfully_imported_settings));
     }
 
-    private Collection<String> getAllKeys() {
-        Collection<String> keys = new ArrayList<>();
-        keys.addAll(ALL_GENERAL_KEYS);
-        keys.add(KEY_ADMIN_PW);
+    private Collection<String> getAllGeneralKeys() {
+        Collection<String> keys = GENERAL_KEYS.keySet();
         keys.add(KEY_PASSWORD);
```

Comment:
`keySet` doesn't support `add`, see [here](http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html#keySet%28%29). I think the right thing to do is to use the copy constructor.